storm/src/main/storm/mesos/MesosNimbus.java

```diff
@@ -577,288 +409,267 @@ public void assignSlots(Topologies topologies, Map<String, Collection<WorkerSlot
+          workerPrefix = MesosCommon.getWorkerPrefix(mesosStormConf, topologyDetails);
         }
-      }
 
-      // The following way of finding resources for the slot is bad!
-      // Suppose we have the following offers:
-      //     o1 - { host: h1 ports : 31000-32000 }
-      //     o2 - { host: h1 mem : 30000 cpu : 24 }
-      // because the offers are fragmented, they are useless
-      //   1. At this point a worker slot "h1-3000" associated with o1 for instance wouldn't
-      //      be launched because it doesn't have mem and cpu
-      //   2. o2 is useless because it doesn't have any ports and therefore won't be used at all
-      if (offerId != null && offer != null) {
-        // The fact that we are here means that this offer has a port. We need to find if it also has
-        // enough memory and cpu to launch the worker.
-        if (!toLaunch.containsKey(offerId)) {
-          toLaunch.put(offerId, new ArrayList<LaunchTask>());
+        if (!hostsWithSupervisors.contains(workerHost)) {
+          requiredCpu += executorCpu;
+          requiredMem += executorMem;
         }
-        double workerCpu = MesosCommon.topologyWorkerCpu(_conf, details);
-        double workerMem = MesosCommon.topologyWorkerMem(_conf, details);
-        double executorCpu = MesosCommon.executorCpu(_conf);
-        double executorMem = MesosCommon.executorMem(_conf);
 
         Map executorData = new HashMap();
-        executorData.put(MesosCommon.SUPERVISOR_ID, MesosCommon.supervisorId(slot.getNodeId(), details.getId()));
+        executorData.put(MesosCommon.SUPERVISOR_ID, MesosCommon.supervisorId(slot.getNodeId(), topologyDetails.getId()));
         executorData.put(MesosCommon.ASSIGNMENT_ID, workerPrefix + slot.getNodeId());
 
-        Offer.Builder newBuilder = Offer.newBuilder();
-        newBuilder.mergeFrom(offer);
-        newBuilder.clearResources();
-
-        Offer.Builder existingBuilder = Offer.newBuilder();
-        existingBuilder.mergeFrom(offer);
-        existingBuilder.clearResources();
+        String topologyAndNodeId = topologyDetails.getId() + " | " + slot.getNodeId();
+        String executorName = "storm-supervisor | " + topologyAndNodeId;
+        String taskName = "storm-worker | " + topologyAndNodeId + ":" + slot.getPort();
+        String executorDataStr = JSONValue.toJSONString(executorData);
         String extraConfig = "";
 
-        List<Resource> offerResources = new ArrayList<>();
-        offerResources.addAll(offer.getResourcesList());
-        // Prefer reserved resources?
-        if (_preferReservedResources) {
-          Collections.sort(offerResources, new ResourceRoleComparator());
+        if (!aggregatedOffers.isFit(mesosStormConf, topologyDetails, workerPort, hostsWithSupervisors.contains(workerHost))) {
+          LOG.error(String.format("Unable to launch worker %s. Required cpu: %f, Required mem: %f. Available aggregatedOffers : %s",
```

Comment:
Since the availability of this particular port will determine whether or not the assignment will fit, can we also include the `workerPort` in this log line?